p2p/protocol/holepunch/coordination.go

```diff
@@ -0,0 +1,388 @@
+package holepunch
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"sync"
+	"sync/atomic"
+	"time"
+
+	pb "github.com/libp2p/go-libp2p/p2p/protocol/holepunch/pb"
+	"github.com/libp2p/go-libp2p/p2p/protocol/identify"
+
+	"github.com/libp2p/go-libp2p-core/host"
+	"github.com/libp2p/go-libp2p-core/network"
+	"github.com/libp2p/go-libp2p-core/peer"
+	"github.com/libp2p/go-libp2p-core/protocol"
+
+	logging "github.com/ipfs/go-log"
+	"github.com/libp2p/go-msgio/protoio"
+	ma "github.com/multiformats/go-multiaddr"
+	manet "github.com/multiformats/go-multiaddr/net"
+)
+
+// Protocol is the libp2p protocol for Hole Punching.
+const Protocol protocol.ID = "/libp2p/dcutr"
+
+// StreamTimeout is the timeout for the hole punch protocol stream.
+var StreamTimeout = 1 * time.Minute
+
+// TODO Should we have options for these ?
+const (
+	maxMsgSize  = 4 * 1024 // 4K
+	dialTimeout = 5 * time.Second
+	maxRetries  = 3
+	retryWait   = 2 * time.Second
+)
+
+var (
+	log = logging.Logger("p2p-holepunch")
+)
+
+// The Service is used to make direct connections with a peer via hole-punching.
+type Service struct {
+	ctx       context.Context
+	ctxCancel context.CancelFunc
+
+	ids  *identify.IDService
+	host host.Host
+
+	tracer *Tracer
+
+	refCount sync.WaitGroup
+	counter  int32 // to be used as an atomic. How many hole punch requests we have responded to
+
+	// active hole punches for deduplicating
+	activeMx sync.Mutex
+	active   map[peer.ID]struct{}
+}
+
+type Option func(*Service) error
+
+// NewService creates a new service that can be used for hole punching
+func NewService(h host.Host, ids *identify.IDService, opts ...Option) (*Service, error) {
+	if ids == nil {
+		return nil, errors.New("identify service can't be nil")
+	}
+
+	ctx, cancel := context.WithCancel(context.Background())
+	hs := &Service{
+		ctx:       ctx,
+		ctxCancel: cancel,
+		host:      h,
+		ids:       ids,
+		active:    make(map[peer.ID]struct{}),
+	}
+
+	for _, opt := range opts {
+		if err := opt(hs); err != nil {
+			cancel()
+			return nil, err
+		}
+	}
+
+	h.SetStreamHandler(Protocol, hs.handleNewStream)
+	h.Network().Notify((*netNotifiee)(hs))
+	return hs, nil
+}
+
+// Close closes the Hole Punch Service.
+func (hs *Service) Close() error {
+	hs.ctxCancel()
+	hs.refCount.Wait()
```

Comment:
This can panic because the count can go from N->0->M. 

I'm not sure if this service actually needs to be closable, honestly. It doesn't do anything in the background and will stop automatically when the host is closed.

But if we do want to do this, I'd protect `wg.Add` with a lock (checking to see if we've been "canceled"). This is also a good use-case for conditions (wait until the number of active jobs has reached zero, and broadcast whenever a job is finished).

Reply:
> I'm not sure if this service actually needs to be closable, honestly. It doesn't do anything in the background

It does. `HolePunch` is called for a separate go routine, and it sleeps for `rtt /2`. Note that `rtt` can be large. We need to be able to cancel that sleeping period.

Reply:
> This is also a good use-case for conditions (wait until the number of active jobs has reached zero, and broadcast whenever a job is finished).

I just used a bool and a mutex, not sure if the condition would actually make things easier here (but maybe I'm just misunderstanding your proposal).